contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,351 @@
+            @gcalendar <event_title> | <start_date> | <end_date> | <timezone> (optional)
+        For full-day events:
+            @gcalendar <event_title> | <start_date>
+
+    The "event-title" supports all characters but pipes (|)
+
+    Timezones can be specified in both numerical (+00:00) or abbreviated
+    format (UTC).
+    If no timezone is specified, the server's will be considered as default.
+
+    Right now it only works for the calendar set as "primary" in your account.
+    '''
+    def __init__(self):
+        credentials = get_credentials()
+        http = credentials.authorize(httplib2.Http())
+        self.service = discovery.build('calendar', 'v3', http=http)
+
+    def usage(self):
+        return '''
+            This plugin will allow users to create events
+            in a Google Calendar. Users should preface
+            messages with "@gcalendar".
+
+            Before running this, make sure to register a new
+            project in the Google Developers Console
+            (https://console.developers.google.com/start/api?id=calendar)
+            and download "client secret" as a JSON file,
+            and save it to ~/.gcal_client_secret.json.
+            '''
+
+    def triage_message(self, message, client):
+        original_content = message['content']
+
+        is_gcalendar = original_content.startswith('@gcalendar')
+
+        return is_gcalendar
+
+    def handle_message(self, message, client, state_handler):
+        if message['content'] == '@gcalendar help':
+            send_help_message(message, client)
+            return
+
+        try:
+            title, start_date, end_date = parse_message(message['content'])
+        except MessageParseError as e: # Something went wrong during parsing
+            send_parsing_error_message(e, message, client)
+            return
+
+        event = {
+            'summary': title
+        }
+
+        if not end_date: # Full-day event
+            date = start_date.strftime('%Y-%m-%d')
+            event.update({
+                'start': {'date': date},
+                'end': {'date': date}
+            })
+        else:
+            if not start_date.tzinfo:
+                start_date = start_date.replace(tzinfo=get_localzone())
+            if not end_date.tzinfo:
+                end_date = end_date.replace(tzinfo=get_localzone())
+
+            event.update({
+                'start': {'dateTime': start_date.isoformat()},
+                'end': {'dateTime': end_date.isoformat()}
+            })
+
+        try:
+            # TODO: Choose calendar ID from somewhere
+            event = self.service.events().insert(calendarId='primary', body=event).execute()
+        except errors.HttpError as e:
+            logging.exception(e)
+
+            error = (':warning: **Error!**\n Something went wrong.\n'
+                     'Please, try again or check the logs if the issue persists.')
+
+            reply_message(error, message, client)
```

Comment:
I added the possibility of `reply_message` raising such an exception just in case someone extended the bot in the future with some other features. Right now, there's no possible way that an `UnknownMessageTypeError` is raised (only if Zulip provided an invalid message type, which would be rather worrying).

Should they be handled anyway? I'm just asking, because it would slightly clutter up the code, since `reply_message` is called multiple times along the bot's code.